src/rust/src/tracing_ffi.rs

```diff
@@ -0,0 +1,552 @@
+use libc::c_char;
+use std::ffi::CStr;
+use std::path::Path;
+use std::slice;
+use std::str;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use tracing::{
+    callsite::{Callsite, Identifier},
+    field::{FieldSet, Value},
+    level_enabled,
+    metadata::Kind,
+    span::Entered,
+    subscriber::{Interest, Subscriber},
+    Event, Metadata, Span,
+};
+use tracing_appender::non_blocking::WorkerGuard;
+use tracing_core::Once;
+use tracing_subscriber::{
+    filter::EnvFilter,
+    layer::Layer,
+    reload::{self, Handle},
+};
+
+#[cfg(not(target_os = "windows"))]
+use std::ffi::OsStr;
+#[cfg(not(target_os = "windows"))]
+use std::os::unix::ffi::OsStrExt;
+
+#[cfg(target_os = "windows")]
+use std::ffi::OsString;
+#[cfg(target_os = "windows")]
+use std::os::windows::ffi::OsStringExt;
+
+trait ReloadHandle {
+    fn reload(&self, new_filter: EnvFilter) -> Result<(), reload::Error>;
+}
+
+impl<L, S> ReloadHandle for Handle<L, S>
+where
+    L: From<EnvFilter> + Layer<S> + 'static,
+    S: Subscriber,
+{
+    fn reload(&self, new_filter: EnvFilter) -> Result<(), reload::Error> {
+        self.reload(new_filter)
+    }
+}
+
+pub struct TracingHandle {
+    _file_guard: Option<WorkerGuard>,
+    reload_handle: Box<dyn ReloadHandle>,
+}
+
+#[no_mangle]
+pub extern "C" fn tracing_init(
+    #[cfg(not(target_os = "windows"))] log_path: *const u8,
+    #[cfg(target_os = "windows")] log_path: *const u16,
+    log_path_len: usize,
+    initial_filter: *const c_char,
+    log_timestamps: bool,
+) -> *mut TracingHandle {
+    let initial_filter = unsafe { CStr::from_ptr(initial_filter) }
+        .to_str()
+        .expect("initial filter should be a valid string");
```

Comment:
Is it okay to use `expect` here? 

My understanding is that [panicking across an FFI boundary is undefined behavior](https://doc.rust-lang.org/nomicon/ffi.html#ffi-and-panics). AFAICT, this is called directly from C++ code, without wrapping it in a `std::panic::catch_unwind`, so if an invalid filter string is provided, this would result in UB.

I think (although I'm not entirely sure; I'm not a Rust FFI expert), it might be more correct to `println` the error and return a null pointer here, or change the function signature to take an `*mut TracingHandle` as an in-parameter and return an error code? Unless there's another more correct way of reporting errors I'm not aware of...